### ISSUE TITLE
File.read_to_end doesn't always read to the end

### ISSUE THREAD

**Issue by joelself** [rust-lang/rust#31488]
```rust
let mut file = File::open("test.log").unwrap();
file.seek(SeekFrom::End(0)).unwrap(); // seek to the end
// elided, wait for file modified signal then read what was written
let bytes_read = file.read_to_end(& mut buffer).unwrap();
match String::from_utf8(buffer.clone()) {
	Ok(s) => println!("Read: {}, bytes: {}", String::from_utf8(buffer.clone()).unwrap(), bytes_read),
	_ => (),
}
```
I elided the code where I wait on a signal that the file has changed. If you do something like
```shell
echo "Hello World!" >> test.log
```
It works fine; gets the bytes and prints it out. However something like
```shell
sed -i.bck '$s/$/GoodByeWorld/' test.log
```
Will modify the file, trigger the file modified signal I have in the code, but it won't read any bytes. This is definitely an unusual case as `tail -f` doesn't catch it either even if you later append normally like: `echo "Hello World!" >> test.log'`. `tail -F` however (after a very brief delay) *will* pick it up.

So the file is being appended to and an event is being generated for the append, but the Rust File still thinks it is already at EOF.

**Comment by tmiasko** [comment]
The in place flag of sed may be a little misleading, it does leave the result in
the same path, but it is a different file. It works by writing output to a new
temporary file and then doing rename (you can confirm this with strace). So to
detect this change you would have to reopen the file, and this is exactly what
tail --follow=name does (implied by -F).

**Comment by joelself** [comment]
Thanks @tmiasko , I googled this around and found out about this. Now when I get a write event, but no bytes read, I automatically reopen the file and it works for the most part as long as I keep track of my position in the file. The one problem I have is when I reopen the file and do a `read_to_end` it always tacks on a '\n' to the end of the buffer even if there shouldn't be one (I can tell because subsequent writes go to the same line). Should I open a different issue for this?
